### ISSUE TITLE
smaller systems

### ISSUE THREAD

**Issue by richardjgowers** [bio-phys/MDBenchmark#12]
disclaimer: I've not had time to poke around too much inside.

This looks like it's made for large systems, but it could also be useful for smaller (single node) systems.  When running on a single node there's a question of how many mpi ranks/openmp threads to use, the answer (probably?) depending on system size and how many particles per task you end up with.

How easy/hard would it be to adapt to do this?

**Comment by kain88-de** [comment]
I'm currently fixed the ratio of mpi ranks / openmp threads to 1/2 (i.e 2 openmp threads per mpi rank).

I thought about allowing to allow changing this as well. This is definitely doable without much effort in code. It's a question how to expose this to a user. 

I would opt that we always have a fixed number of available logical cores `n_cores` and that the number of ranks `n_ranks` and openmp threads `n_threads` should be constraint by `n_cores = n_ranks * n_threads`.   Then on the cli we give a partition parameter with `--partition = 2`  as a default. If one then wants to change the partition you can use for example `--partition = 3` which means that we use 3 openmp threads per mpi rank. Of course then we should also allow specifying the number of cores to use in `mdbenchmark generate`. Does that sound good to you?

Out folder naming would then also update. We need a combination of `n_nodes-n_partition`.

**Comment by richardjgowers** [comment]
it might make more sense to allow a list of partition numbers then, `--partition = 1,2,4,8,16`.

And yeah maybe `n_cores` **or** `n_nodes`, (nobody cares about 1.5 nodes I guess)

I guess the command I'd want to run is:

`mdbenchmark generate --name protein --max-cores 16 --partition 1,2,4,8,16`

Which would produce a 2d plot of xaxis=n_cores, yaxis=partition, colour=performance.  Ie how much hardware can I throw at my tiny system before I'm wasting my resources.

**Comment by kain88-de** [comment]
Here for me `n_cores` always refers to the number of cores on a single node!  So I need both.

What you want to optimize is what is my performance on a single node when I increase the number of cores and change the partitioning. So you want a 2D graph like this.

```
 n_cores |
         |
         |________
                  partition
```

Color would indicate performance. 

But for larger system we get a third variable the number of nodes. So I will have to think a bit about how to expose this in the UI.

Also some guidelines and restrictions are good then because this is a brute-force search in parameter space for optimal performance which can explode the number of started simulations very quickly.

**Comment by kain88-de** [comment]
@richardjgowers do you run GROMACS with this or do you plan to use another tool? The pure benchmark setup and execution is pretty agnostic to gromacs when you write your own host-template files. Only the `analyze` function currently expects that you run a gromacs simulation.

I'm asking because we have noticed that newer versions of gromacs do some internal performance tuning and overwrite your settings when run on a single node.

**Comment by richardjgowers** [comment]
I figured once you're above a single node you can assume that you're using all available cores on a node - so really all I'm asking is for this tool to allow sizes under a single node (granularity of cores when "under" a single node in size).

I'm asking for a friend mostly, so if Gromacs is smart enough to do the partition tuning itself, maybe this is a non-issue.

**Comment by kain88-de** [comment]
http://manual.gromacs.org/documentation/2016.4/user-guide/mdrun-performance.html#running-mdrun-within-a-single-node

> By default, gmx mdrun will inspect the hardware available at run time and do its best to make fairly efficient use of the whole node. The log file, stdout and stderr are used to print diagnostics that inform the user about the choices made and possible consequences.

I think you can still set the number of MPI ranks used per node as gromacs can't really change that. We could add this if it helps and he really can run effectively several small systems on a single node.  I'm not sure how useful this is to do though.

**Comment by kain88-de** [comment]
@richardjgowers does this work for your friend. I'm not opposed to adding more options but I have never used systems so small that I could run several of them on a single node.

**Comment by kain88-de** [comment]
@richardjgowers Here is a talk about configuring gromacs for performance.

https://www.slideshare.net/BioExcel/bioexcel-webinar-series-2-performance-tuning-and-optimization-of-gromacs

It also talks about single node options with the omp threads vs mpi that one can try. I don't have much experience with that yet. I'm thinking of adding these to the test matrix that can be generated.

**Comment by mimischi** [comment]
Hey @richardjgowers,

this is now implemented in the `version-3` branch via PR #165. I am closing this old issue for now, but feel free to re-open if necessary. The PR has some hints how to use it, documentation is following soon.
